### PR TITLE
MMR add bagging and merkle root

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description
This PR adds support for bagging peaks, and add a merkle root function

## Motivation and Context
Expands the MMR code to add all basic needed functions

## How Has This Been Tested?
The code effects the MMR a bit, all unit tests were rerun and all passed. The unit tests were also expanded to include more edge cases.

## Types of changes
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by SWvheerden** [comment]
fixes #126

File: infrastructure/merklemountainrange/src/merklemountainrange.rs [hunk 1]
```diff
@@ -159,13 +236,36 @@ where
...
 
     // This function is just a private function to return the index of the last added node
     fn get_last_added_index(&self) -> usize {
         self.mmr.len() - 1
     }
+
+    fn bag_mmr(&self) -> Vec<ObjectHash> {
+        // lets find all peaks of the mmr
+        let mut peaks = Vec::new();
+        self.find_bagging_indexes(self.current_peak_height.0 as i64, self.current_peak_height.1, &mut peaks);
```
Comment (philipr-za) [inline infrastructure/merklemountainrange/src/merklemountainrange.rs]: Feels a little clunky using a mutable reference to act as a return variable, why not just have `find_bagging_indexes(...) -> Vec<ObjectHash>`. That makes it easier to see what the function is actually producing.

Comment (SWvheerden) [reply infrastructure/merklemountainrange/src/merklemountainrange.rs]: Look down below, its recursive function. 
Either you need to clone/copy/move the vec each time the function is called, or you just pass in the same vec and keep appending to it.

Comment (philipr-za) [reply infrastructure/merklemountainrange/src/merklemountainrange.rs]: Oh, ok
